File: src/lexer/lexer.py
#!/usr/bin/env python3
"""
Lexer para a linguagem de programação Aqua
Analisa o código fonte e gera tokens para o parser
"""

from enum import Enum, auto
from typing import List, Optional, Tuple
import re

class TokenType(Enum):
    # Palavras-chave
    FUNC = auto()
    LET = auto()
    IMPORT = auto()
    SPAWN = auto()
    LOOP = auto()
    MATCH = auto()
    CASE = auto()
    BREAK = auto()
    RETURN = auto()
    
    # Identificadores e literais
    IDENTIFIER = auto()
    STRING = auto()
    NUMBER = auto()
    BOOLEAN = auto()
    
    # Operadores
    PLUS = auto()
    MINUS = auto()
    MULTIPLY = auto()
    DIVIDE = auto()
    ASSIGN = auto()
    EQUALS = auto()
    NOT_EQUALS = auto()
    GREATER = auto()
    LESS = auto()
    GREATER_EQUAL = auto()
    LESS_EQUAL = auto()
    
    # Delimitadores
    LPAREN = auto()
    RPAREN = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    COMMA = auto()
    SEMICOLON = auto()
    COLON = auto()
    ARROW = auto()
    DOT = auto()
    
    # Especiais
    NEWLINE = auto()
    INDENT = auto()
    DEDENT = auto()
    EOF = auto()

class Token:
    def __init__(self, type: TokenType, value: str, line: int, column: int):
        self.type = type
        self.value = value
        self.line = line
        self.column = column
    
    def __repr__(self):
        return f"Token({self.type}, '{self.value}', line={self.line}, col={self.column})"

class Lexer:
    def __init__(self, source: str):
        self.source = source
        self.position = 0
        self.line = 1
        self.column = 1
        self.indent_stack = [0]
        self.tokens = []
        
        # Padrões para tokens
        self.patterns = [
            # Palavras-chave
            (r'\bfunc\b', TokenType.FUNC),
            (r'\blet\b', TokenType.LET),
            (r'\bimport\b', TokenType.IMPORT),
            (r'\bspawn\b', TokenType.SPAWN),
            (r'\bloop\b', TokenType.LOOP),
            (r'\bmatch\b', TokenType.MATCH),
            (r'\bcase\b', TokenType.CASE),
            (r'\bbreak\b', TokenType.BREAK),
            (r'\breturn\b', TokenType.RETURN),
            (r'\btrue\b', TokenType.BOOLEAN),
            (r'\bfalse\b', TokenType.BOOLEAN),
            
            # Números
            (r'\d+\.\d+', TokenType.NUMBER),  # Float
            (r'\d+', TokenType.NUMBER),        # Integer
            
            # Strings
            (r'"[^"]*"', TokenType.STRING),
            (r"'[^']*'", TokenType.STRING),
            
            # Operadores
            (r'\+', TokenType.PLUS),
            (r'-', TokenType.MINUS),
            (r'\*', TokenType.MULTIPLY),
            (r'/', TokenType.DIVIDE),
            (r'=', TokenType.ASSIGN),
            (r'==', TokenType.EQUALS),
            (r'!=', TokenType.NOT_EQUALS),
            (r'>', TokenType.GREATER),
            (r'<', TokenType.LESS),
            (r'>=', TokenType.GREATER_EQUAL),
            (r'<=', TokenType.LESS_EQUAL),
            (r'=>', TokenType.ARROW),
            
            # Delimitadores
            (r'\(', TokenType.LPAREN),
            (r'\)', TokenType.RPAREN),
            (r'\[', TokenType.LBRACKET),
            (r'\]', TokenType.RBRACKET),
            (r',', TokenType.COMMA),
            (r';', TokenType.SEMICOLON),
            (r':', TokenType.COLON),
            (r'\.', TokenType.DOT),
            
            # Identificadores
            (r'[a-zA-Z_][a-zA-Z0-9_]*', TokenType.IDENTIFIER),
        ]
        
        # Compilar padrões regex
        self.compiled_patterns = [(re.compile(pattern), token_type) 
                                 for pattern, token_type in self.patterns]
    
    def tokenize(self) -> List[Token]:
        """Analisa o código fonte e retorna uma lista de tokens"""
        while self.position < len(self.source):
            # Pular espaços em branco
            if self.skip_whitespace():
                continue
            
            # Pular comentários
            if self.skip_comments():
                continue
            
            # Processar nova linha e indentação
            if self.source[self.position] == '\n':
                self.handle_newline()
                continue
            
            # Tentar fazer match com padrões
            token = self.match_patterns()
            if token:
                self.tokens.append(token)
                continue
            
            # Caractere não reconhecido
            raise SyntaxError(f"Caractere não reconhecido '{self.source[self.position]}' "
                           f"na linha {self.line}, coluna {self.column}")
        
        # Adicionar tokens de indentação finais
        self.handle_final_indentation()
        
        # Adicionar EOF
        self.tokens.append(Token(TokenType.EOF, "", self.line, self.column))
        
        return self.tokens
    
    def skip_whitespace(self) -> bool:
        """Pula espaços em branco e retorna True se pulou algum"""
        skipped = False
        while (self.position < len(self.source) and 
               self.source[self.position].isspace() and 
               self.source[self.position] != '\n'):
            self.position += 1
            self.column += 1
            skipped = True
        return skipped
    
    def skip_comments(self) -> bool:
        """Pula comentários e retorna True se pulou algum"""
        if (self.position < len(self.source) and 
            self.source[self.position] == '#'):
            # Comentário de linha única
            while (self.position < len(self.source) and 
                   self.source[self.position] != '\n'):
                self.position += 1
                self.column += 1
            return True
        return False
    
    def handle_newline(self):
        """Processa uma nova linha e gerencia indentação"""
        self.tokens.append(Token(TokenType.NEWLINE, "\n", self.line, self.column))
        self.position += 1
        self.line += 1
        self.column = 1
        
        # Calcular indentação da próxima linha
        indent_level = 0
        while (self.position < len(self.source) and 
               self.source[self.position] == ' '):
            indent_level += 1
            self.position += 1
            self.column += 1
        
        # Gerar tokens de indentação
        current_indent = self.indent_stack[-1]
        if indent_level > current_indent:
            self.indent_stack.append(indent_level)
            self.tokens.append(Token(TokenType.INDENT, "", self.line, self.column))
        elif indent_level < current_indent:
            while self.indent_stack and self.indent_stack[-1] > indent_level:
                self.indent_stack.pop()
                self.tokens.append(Token(TokenType.DEDENT, "", self.line, self.column))
    
    def match_patterns(self) -> Optional[Token]:
        """Tenta fazer match com os padrões de tokens"""
        for pattern, token_type in self.compiled_patterns:
            match = pattern.match(self.source, self.position)
            if match:
                value = match.group(0)
                token = Token(token_type, value, self.line, self.column)
                self.position += len(value)
                self.column += len(value)
                return token
        return None
    
    def handle_final_indentation(self):
        """Adiciona tokens de dedentação finais"""
        while len(self.indent_stack) > 1:
            self.indent_stack.pop()
            self.tokens.append(Token(TokenType.DEDENT, "", self.line, self.column))

def lex(source: str) -> List[Token]:
    """Função conveniente para tokenizar código fonte"""
    lexer = Lexer(source)
    return lexer.tokenize()

if __name__ == "__main__":
    # Exemplo de uso
    source_code = '''
func main()
    let x = 42
    let message = "Olá, Aqua!"
    io.println(message)
    io.println(x)
'''
    
    tokens = lex(source_code)
    for token in tokens:
        print(token)

File: src/lexer/lexer.cpp
#include "lexer.hpp"
#include "utils/error.hpp"

#include <cctype>
#include <sstream>
#include <unordered_map>

namespace aqua::lexing {

static const std::unordered_set<std::string> kKeywords = {
    "func","let","import","spawn","match","case","loop","break","continue",
    "if","else","return","make_channel","sleep","true","false","None",
    // tipos
    "int","float","string","bool"
};

Lexer::Lexer(std::string source)
    : input(std::move(source)) {}

bool Lexer::isKeyword(std::string_view word) {
    return kKeywords.find(std::string(word)) != kKeywords.end();
}

char Lexer::peek(size_t lookahead) const {
    size_t pos = index + lookahead;
    if (pos >= input.size()) return '\0';
    return input[pos];
}

bool Lexer::eof(size_t lookahead) const {
    return index + lookahead >= input.size();
}

char Lexer::advance() {
    if (eof()) return '\0';
    char c = input[index++];
    if (c == '\n') {
        line++;
        column = 1;
        atLineStart = true;
    } else {
        column++;
    }
    return c;
}

void Lexer::emit(TokenType type, const std::string& value) {
    tokens.push_back(Token{type, value, line, column});
}

void Lexer::skipSpaces() {
    while (!eof() && peek() == ' ' && !atLineStart) {
        advance();
    }
}

bool Lexer::skipComment() {
    if (!eof() && peek() == '#') {
        while (!eof() && advance() != '\n') { /* skip */ }
        return true;
    }
    return false;
}

void Lexer::emitPendingIndents(int indent) {
    int current = indentStack.back();
    if (indent > current) {
        indentStack.push_back(indent);
        tokens.push_back(Token{TokenType::INDENT, "", line, column});
    } else if (indent < current) {
        while (!indentStack.empty() && indentStack.back() > indent) {
            indentStack.pop_back();
            tokens.push_back(Token{TokenType::DEDENT, "", line, column});
        }
        if (indentStack.empty() || indentStack.back() != indent) {
            throw LexerError(formatError("IndentationError", line, column, "Indentação inválida"));
        }
    }
}

void Lexer::handleNewline() {
    // Consumir o \n
    advance();
    tokens.push_back(Token{TokenType::NEWLINE, "\n", line - 1, 1});

    // Calcular indentação da próxima linha: apenas espaços; tabs não suportados
    int indent = 0;
    while (!eof() && peek() == ' ') {
        indent++;
        advance();
    }
    atLineStart = false; // já consumimos espaços do início

    // Se a linha está vazia ou comentário, mantenha estado de início de linha para próxima
    if (peek() == '\n' || peek() == '#') {
        // linhas vazias não alteram indentação
        return;
    }

    emitPendingIndents(indent);
}

static bool isDecimalDigit(char c) { return std::isdigit(static_cast<unsigned char>(c)) != 0; }

std::optional<Token> Lexer::readNumber() {
    if (!isDecimalDigit(peek())) return std::nullopt;

    int startLine = line;
    int startColumn = column;

    std::string lexeme;
    bool hasDot = false;

    while (!eof()) {
        char c = peek();
        if (isDecimalDigit(c)) {
            lexeme.push_back(advance());
        } else if (c == '.' && !hasDot && isDecimalDigit(peek(1))) {
            hasDot = true;
            lexeme.push_back(advance());
        } else {
            break;
        }
    }

    return Token{TokenType::NUMBER, lexeme, startLine, startColumn};
}

bool Lexer::isIdentStart(char c) {
    return std::isalpha(static_cast<unsigned char>(c)) || c == '_';
}

bool Lexer::isIdentPart(char c) {
    return std::isalnum(static_cast<unsigned char>(c)) || c == '_';
}

std::optional<Token> Lexer::readIdentifierOrKeyword() {
    if (!isIdentStart(peek())) return std::nullopt;

    int startLine = line;
    int startColumn = column;

    std::string lexeme;
    while (!eof() && isIdentPart(peek())) {
        lexeme.push_back(advance());
    }

    // keywords e booleanos/None
    if (lexeme == "true") return Token{TokenType::TRUE, lexeme, startLine, startColumn};
    if (lexeme == "false") return Token{TokenType::FALSE, lexeme, startLine, startColumn};
    if (lexeme == "None") return Token{TokenType::NONE, lexeme, startLine, startColumn};

    if (lexeme == "func") return Token{TokenType::FUNC, lexeme, startLine, startColumn};
    if (lexeme == "let") return Token{TokenType::LET, lexeme, startLine, startColumn};
    if (lexeme == "import") return Token{TokenType::IMPORT, lexeme, startLine, startColumn};
    if (lexeme == "spawn") return Token{TokenType::SPAWN, lexeme, startLine, startColumn};
    if (lexeme == "match") return Token{TokenType::MATCH, lexeme, startLine, startColumn};
    if (lexeme == "case") return Token{TokenType::CASE, lexeme, startLine, startColumn};
    if (lexeme == "loop") return Token{TokenType::LOOP, lexeme, startLine, startColumn};
    if (lexeme == "break") return Token{TokenType::BREAK, lexeme, startLine, startColumn};
    if (lexeme == "continue") return Token{TokenType::CONTINUE, lexeme, startLine, startColumn};
    if (lexeme == "if") return Token{TokenType::IF, lexeme, startLine, startColumn};
    if (lexeme == "else") return Token{TokenType::ELSE, lexeme, startLine, startColumn};
    if (lexeme == "return") return Token{TokenType::RETURN, lexeme, startLine, startColumn};
    if (lexeme == "make_channel") return Token{TokenType::MAKE_CHANNEL, lexeme, startLine, startColumn};
    if (lexeme == "sleep") return Token{TokenType::SLEEP, lexeme, startLine, startColumn};

    if (lexeme == "int") return Token{TokenType::INT, lexeme, startLine, startColumn};
    if (lexeme == "float") return Token{TokenType::FLOAT, lexeme, startLine, startColumn};
    if (lexeme == "string") return Token{TokenType::STRING_TYPE, lexeme, startLine, startColumn};
    if (lexeme == "bool") return Token{TokenType::BOOL, lexeme, startLine, startColumn};
    if (lexeme == "and") return Token{TokenType::AND, lexeme, startLine, startColumn};
    if (lexeme == "or") return Token{TokenType::OR, lexeme, startLine, startColumn};
    if (lexeme == "not") return Token{TokenType::NOT, lexeme, startLine, startColumn};

    return Token{TokenType::IDENTIFIER, lexeme, startLine, startColumn};
}

std::optional<Token> Lexer::readString() {
    if (peek() != '"') return std::nullopt;

    int startLine = line;
    int startColumn = column;

    advance(); // opening quote
    std::string result;

    while (!eof()) {
        char c = advance();
        if (c == '"') {
            return Token{TokenType::STRING, result, startLine, startColumn};
        }
        if (c == '\\') {
            if (eof()) break;
            char esc = advance();
            switch (esc) {
                case 'n': result.push_back('\n'); break;
                case 't': result.push_back('\t'); break;
                case '"': result.push_back('"'); break;
                case '\\': result.push_back('\\'); break;
                default:
                    // mantém literal desconhecido como é
                    result.push_back('\\');
                    result.push_back(esc);
                    break;
            }
        } else {
            result.push_back(c);
        }
    }

    throw LexerError(formatError("LexError", startLine, startColumn, "String não terminada"));
}

std::optional<Token> Lexer::readOperatorOrPunct() {
    int startLine = line;
    int startColumn = column;

    char c = peek();

    // Operadores compostos
    if (c == ':' && peek(1) == '=') {
        advance(); advance();
        return Token{TokenType::COLON_EQ, ":=", startLine, startColumn};
    }
    if (c == '=' && peek(1) == '=') { advance(); advance(); return Token{TokenType::EQEQ, "==", startLine, startColumn}; }
    if (c == '!' && peek(1) == '=') { advance(); advance(); return Token{TokenType::BANGEQ, "!=", startLine, startColumn}; }
    if (c == '>' && peek(1) == '=') { advance(); advance(); return Token{TokenType::GTE, ">=", startLine, startColumn}; }
    if (c == '<' && peek(1) == '=') { advance(); advance(); return Token{TokenType::LTE, "<=", startLine, startColumn}; }
    if (c == '=' && peek(1) == '>') { advance(); advance(); return Token{TokenType::ARROW, "=>", startLine, startColumn}; }
    if (c == '-' && peek(1) == '>') { advance(); advance(); return Token{TokenType::RARROW, "->", startLine, startColumn}; }

    // Simples
    switch (c) {
        case '+': advance(); return Token{TokenType::PLUS, "+", startLine, startColumn};
        case '-': advance(); return Token{TokenType::MINUS, "-", startLine, startColumn};
        case '*': advance(); return Token{TokenType::STAR, "*", startLine, startColumn};
        case '/': advance(); return Token{TokenType::SLASH, "/", startLine, startColumn};
        case '%': advance(); return Token{TokenType::PERCENT, "%", startLine, startColumn};
        case '=': advance(); return Token{TokenType::EQ, "=", startLine, startColumn};
        case '>': advance(); return Token{TokenType::GT, ">", startLine, startColumn};
        case '<': advance(); return Token{TokenType::LT, "<", startLine, startColumn};
        case '(': advance(); return Token{TokenType::LPAREN, "(", startLine, startColumn};
        case ')': advance(); return Token{TokenType::RPAREN, ")", startLine, startColumn};
        case '[': advance(); return Token{TokenType::LBRACKET, "[", startLine, startColumn};
        case ']': advance(); return Token{TokenType::RBRACKET, "]", startLine, startColumn};
        case '{': advance(); return Token{TokenType::LBRACE, "{", startLine, startColumn};
        case '}': advance(); return Token{TokenType::RBRACE, "}", startLine, startColumn};
        case ',': advance(); return Token{TokenType::COMMA, ",", startLine, startColumn};
        case '.': advance(); return Token{TokenType::DOT, ".", startLine, startColumn};
        case ':': advance(); return Token{TokenType::COLON, ":", startLine, startColumn};
        default:
            return std::nullopt;
    }
}

std::vector<Token> Lexer::tokenize() {
    tokens.clear();

    // Log simples
    // std::cerr << "[Lexer] Iniciando tokenização..." << std::endl;

    while (!eof()) {
        // Início de linha: processar indentação
        if (atLineStart) {
            int indent = 0;
            while (!eof() && peek() == ' ') { indent++; advance(); }
            atLineStart = false;

            // Linha vazia ou comentário: emite NEWLINE e continua
            if (peek() == '\n' || peek() == '#') {
                // não altera indentação em linhas vazias
                // consome comentário se houver
                if (peek() == '#') { while (!eof() && advance() != '\n') {} }
                // consome a quebra (se presente)
                if (peek() == '\n') { advance(); }
                tokens.push_back(Token{TokenType::NEWLINE, "\n", line - 1, 1});
                continue;
            }

            emitPendingIndents(indent);
        }

        // Comentários no meio da linha
        if (skipComment()) {
            // Comentário consome até \n; o avanço já contabilizou NEWLINE
            continue;
        }

        // Quebra de linha explícita
        if (peek() == '\n') {
            handleNewline();
            continue;
        }

        // Espaços intermediários (não no início)
        if (peek() == ' ') {
            skipSpaces();
            continue;
        }

        // Strings
        if (auto tok = readString()) { tokens.push_back(*tok); continue; }
        // Número
        if (auto tok = readNumber()) { tokens.push_back(*tok); continue; }
        // Ident/keyword
        if (auto tok = readIdentifierOrKeyword()) { tokens.push_back(*tok); continue; }
        // Operador/pontuação
        if (auto tok = readOperatorOrPunct()) { tokens.push_back(*tok); continue; }

        // Caractere não reconhecido
        std::string msg = std::string("Caractere não reconhecido '") + peek() + "'";
        throw LexerError(formatError("LexError", line, column, msg));
    }

    // Encerrar com dedents pendentes
    while (indentStack.size() > 1) {
        indentStack.pop_back();
        tokens.push_back(Token{TokenType::DEDENT, "", line, column});
    }

    tokens.push_back(Token{TokenType::END_OF_FILE, "", line, column});
    return tokens;
}

} // namespace aqua::lexing
